### PR TITLE
Disallow __arglist in local functions

### LINKED ISSUE #10765: Block __arglist in local functions
Currently, `__arglist` is allowed in the parameter list of local functions, but it is not allowed in the body. This means that the only legal use of `__arglist` in a local function is to declare a method which takes but never uses the `__arglist` parameter.

We should disallow `__arglist` entirely, both from the parameter list and the method body.

### PR DESCRIPTION
**Pull request by agocke**
**Customer scenario**

Currently declaring a local function with an __arglist parameter is not illegal, but it should be.
Disallowing this later is a breaking change.

**Bugs this fixes:** 

Fixes #10765

**Workarounds, if any**

None.

**Risk**

This is a very small change that's specific to local functions. The change is well understood -- it is the same as the check done in lambdas. This change also only prohibits functionality.

**Performance impact**

None.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

New feature

**How was the bug found?**

Customer reported.

### CODE REVIEW

**Comment by agocke** [comment]
@dotnet/roslyn-compiler For review

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 1]
```diff
@@ -448,6 +448,14 @@ private BoundStatement BindLocalFunctionStatement(LocalFunctionStatementSyntax n
             var hasErrors = localSymbol.ScopeBinder
                 .ValidateDeclarationNameConflictsInScope(localSymbol, diagnostics);
 
+            foreach (var p in node.ParameterList.Parameters)
+            {
+                if (p.IsArgList)
+                {
+                    diagnostics.Add(ErrorCode.ERR_IllegalVarArgs, p.Location);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: >diagnostics.Add(ErrorCode.ERR_IllegalVarArgs, p.Location); [](start = 20, length = 58)

It feels like this check belongs in GrabDiagnostics/GetDeclarationDiagnostics on the local function symbol.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs [hunk 2]
```diff
@@ -1244,18 +1244,24 @@ static void Main()
...
+                Diagnostic(ErrorCode.ERR_ArgsInvalid, "__arglist").WithLocation(10, 31),
+                // (18,31): error CS4013: Instance of type 'RuntimeArgumentHandle' cannot be used inside an anonymous function, query expression, iterator block or async method
+                //             Console.WriteLine(__arglist);
+                Diagnostic(ErrorCode.ERR_SpecialByRefInLambda, "__arglist").WithArguments("System.RuntimeArgumentHandle").WithLocation(18, 31),
+                // (24,20): error CS1669: __arglist is not valid in this context
+                //         void Local(__arglist)
+                Diagnostic(ErrorCode.ERR_IllegalVarArgs, "__arglist").WithLocation(24, 20),
+                // (26,31): error CS0190: The __arglist construct is valid only within a variable argument method
+                //             Console.WriteLine(__arglist);
+                Diagnostic(ErrorCode.ERR_ArgsInvalid, "__arglist").WithLocation(26, 31),
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs]: This error feels unexpected, it looks like Binder.BindArgList skips through local functions, but it probably shouldn't.

**Comment by AlekseyTs** [comment]
Is there a reason to disallow __arglist in local functions from the language design point of view? It looks like a simple change in Binder.BindArgList might enable the feature.

**Comment by jaredpar** [comment]
@AlekseyTs the design called for it to be disallowed, mostly because we didn't have a good use case for it. Should keep it disabled at this point as enabling it won't meet the bar.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs (hunk 1 above)
Comment (agocke) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: I was going to eagerly report diagnostics that don't have any risk of cycles and only put lazy things into GetDeclarationDiagnostics. Would you prefer that everything be reported in GetDeclarationDiagnostics if feasible?

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs]: I think we should follow approach when diagnostics about declaration is produced by symbols.

**Comment by AlekseyTs** [comment]
> enabling it won't meet the bar.

Makes sense.

File: src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs (hunk 2 above)
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs]: @AlekseyTs This seems weird for lambdas too... I don't think we should care if the containing member is/is not __arglist if we're inside a lambda/local function while using __arglist. It's invalid no matter what.

I've filed https://github.com/dotnet/roslyn/issues/16720 to rationalize this.

**Comment by agocke** [comment]
test windows_release_unit64_prtest please

**Comment by agocke** [comment]
@MattGertz For escrow approval

File: src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs (hunk 2 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/LocalFunctionTests.cs]: I don't think this is what the logic is about in the place where ERR_ArgsInvalid is reported. Try to add  __arglist to the parameter list of the containing function and see if that makes any difference.

**Comment by natidea** [comment]
Approved to merge via [369813](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/369813)

**Pull request merged**